### PR TITLE
Add Clear Containers 3.0 support

### PR DESCRIPTION
**Pull request by jcvenegas**
This PR adds support for:

### Use docker to run osbuilder scripts is optional:

Now is optional use docker to build the kernel and image.
Use docker allows to build and image and kernel without install all the dependencies (rpm,gcc,bc,losetup).
If docker is wanted to use osbuilder scripts export the variable `USE_DOCKER=true`
```
USE_DOCKER=true
make rootfs
make image
make kernel-src
make kernel
```
ae26e6f osbuilder: allow to build images without docker
0198c4e Makefile: add check-deps target
8e0df1c osbuilder: check if yum or dnf is installed
6b7f2fd image-builder: Use a local mount directory.
de69dfc osbuilder: generate dnf config file in cwd

Fixes: #12

### Check DEBUG env variable to show bash traces.

0291fc8 build: Add DEBUG env variable for osbuilder scripts
Fixes: #13

### Reduce default image size to 50MB
8456741 osbuilder: Only use coreutils-bin
79c0a76 image-builder: Reduce default image size to 50MB
Fixes: #14

### Use URL env variable to change default clearlinux rpm repository.
ba666f0 osbuilder: add env variable to change URL repository
Fixes: #15

### add support for ci scripts to tests in semaphoreci

7059bc5 ci: add semaphoreci setup
eb2822a ci: Run clear containers integration tests

Fixes: #16

### Update kernel to pull the latest version doing `make kernel-src`
4b17f86 kernel: pull latest kernel.
Fixes: #2 

### Clear containers image must inlcude clear-containers-agent
18fa103 pakages: Install clear-containers-agent
Fixes: #17

### CODE REVIEW

**Comment by jcvenegas** [comment]
https://semaphoreci.com/jcvenega/osbuilder/branches/cc30/builds/5 PASS

**Comment by jcvenegas** [comment]
@mcastelino @eadamsintel @jodh-intel  could you please take a look to this PR. ?

File: scripts/image_builder.sh [hunk 1]
```diff
@@ -67,16 +69,16 @@ DEVICE=$(losetup -P -f --show ${IMAGE})
 #Refresh partition table
 partprobe ${DEVICE}
 
-mkdir -p /img
+mkdir -p ./img
```
Comment (jodh-intel) [inline scripts/image_builder.sh:72]: Err - good catch!

File: .ci/ci-setup.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+set -e
+set -x
+
+export test_repo="github.com/clearcontainers/tests"
+export test_repo_dir="${GOPATH}/src/${test_repo}"
```
Comment (jodh-intel) [inline .ci/ci-setup.sh]: These 2 vars are also set in `.ci/ci-run.sh` so how about factoring them out into a `.ci/ci-common.sh` that the other scripts could source?

File: scripts/kernel_builder.sh [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+#  along with this program; if not, write to the Free Software
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+set -e
+
+if [ -n "$DEBUG" ] ; then
+	set -x
+fi
+
+KERNEL_REPO="https://github.com/clearcontainers/linux.git"
```
Comment (jodh-intel) [inline scripts/kernel_builder.sh]: This is also defined in `Makefile`. Can we define it just once somehow?

File: scripts/kernel_builder.sh [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	echo >&2 -e "\e[1mERROR\e[0m: $*"
+	exit 1
+}
+
+info() {
+	echo -e "\e[1mINFO\e[0m: $*"
+}
+
+usage() { 
+	echo "Usage: $0 <subcommand>" 1>&2;
```
Comment (jodh-intel) [inline scripts/kernel_builder.sh]: You could simplify this by doing:

```
cat <<EOT >&2
Options:

-h             : show this help
-k <git-repo>  : git repository to pull linux kernel

    :

EOT
exit 1
```

File: scripts/kernel_builder.sh [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	echo "" 1>&2;
+	echo "Subcommands:" 1>&2;
+	echo "prepare : clone linux source code and patch it with clear-containers configuration" 1>&2;
+	echo "build   : build a kernel configured by prepare subcommand" 1>&2;
+	exit 1; 
+}                                                                                                 
+
+get_last_version() {
+	#Clear Containers kernel release is tagged with \-\d+.container and is incremental
+local cc_release=$(git ls-remote --tags 2>/dev/null | grep -oP '\-\d+\.container'  | grep -oP '\d+' | sort -n | tail -1 )
```
Comment (jodh-intel) [inline scripts/kernel_builder.sh]: All these `git` lines are quite long so I'd be inclined to fold them with some `\`'s.

File: .ci/ci-setup.sh (hunk 2 above)
Comment (jcvenegas) [reply .ci/ci-setup.sh]: Sure, added ci-common.sh :D

File: scripts/kernel_builder.sh (hunk 3 above)
Comment (jcvenegas) [reply scripts/kernel_builder.sh]: Defined only in one place.

File: scripts/kernel_builder.sh (hunk 4 above)
Comment (jcvenegas) [reply scripts/kernel_builder.sh]: Agree, changed.

**Comment by jcvenegas** [comment]
@jodh-intel  thanks for the review, changes applied.

File: Makefile [hunk 6]
```diff
@@ -25,52 +27,89 @@ RUN_PROXY += --env https_proxy=$(https_proxy)
...
-	cd $(WORKDIR); \
-	git clone --depth=1 $(KERNEL_REPO); \
-	cd linux; \
-	make clear_containers_defconfig;
 
 $(WORKDIR):
 	mkdir -p $(WORKDIR)
 
+define check_program
+    ( printf "check for $1..." && type $1 >/dev/null 2>&1 && echo "yes" ) || ( echo "no" && false )
```
Comment (jodh-intel) [inline Makefile:76]: Cute! ;)

File: README.md [hunk 7]
```diff
@@ -59,6 +59,8 @@ workdir/vmlinux.container
 
 
 ```
+#Pull and setup latest kernel for Clear Containers
+sudo -E make kernel-src
```
Comment (jodh-intel) [inline README.md:63]: I would say that this needs a `$ ` at the start. But your change is consistent with the rest of the document so I think adding a prompt can be handled on a separate PR.

File: scripts/kernel_builder.sh [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+#  along with this program; if not, write to the Free Software
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+set -e
+
+if [ -n "$DEBUG" ] ; then
+	set -x
+fi
+
+kernel_repo="https://github.com/clearcontainers/linux.git"
```
Comment (jodh-intel) [inline scripts/kernel_builder.sh:25]: Does this also need to source `.ci/ci-common.sh`?

Comment (jcvenegas) [reply scripts/kernel_builder.sh:25]: Not really, this script it is not part of the ci

File: README.md (hunk 7 above)
Comment (jcvenegas) [reply README.md:63]: Sure, added https://github.com/clearcontainers/osbuilder/issues/20

**Comment by jcvenegas** [comment]
@jodh-intel  @mcastelino  any other comment ?

File: .ci/ci-setup.sh [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+pushd "${test_repo_dir}"
+sudo -E PATH=$PATH bash .ci/setup.sh
+popd
+
+#Vefiry Clear Containers are working before install new image and kernel.
+docker run --rm -ti busybox echo "test" | grep "test"
+
+#Install new image
+#sudo ln -sf "$(pwd)/workdir/vmlinuz.container" /usr/share/clear-containers/vmlinuz.container
+#sudo ln -sf "$(pwd)/workdir/vmlinux.container" /usr/share/clear-containers/vmlinux.container
```
Comment (jodh-intel) [inline .ci/ci-setup.sh]: Ah - just spotted these commented out lines.

**Comment by jodh-intel** [comment]
lgtm

**Pull request merged**